### PR TITLE
cmd: add buy command

### PR DESCRIPTION
**Pull request by pete-woods**

### CODE REVIEW

**Comment by chipaca** [comment]
:+1:

File: cmd/snap/cmd_buy.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		Query: fmt.Sprintf("name:%s", opts.SnapName),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if len(snaps) < 1 {
+		return fmt.Errorf(i18n.G("cannot buy snap %q: it cannot be found"), opts.SnapName)
+	}
```
Comment (chipaca) [inline cmd/snap/cmd_buy.go]: if `len(snaps)>1` then something funny is going on and you should bail.

You probably also want to check `Snapname` doesn't have any of `:*` to provide  a nicer error message

Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Done!

File: cmd/snap/cmd_buy.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/store"
+
+	"github.com/jessevdk/go-flags"
+)
+
+var shortBuyHelp = i18n.G("Finds packages to install")
+var longBuyHelp = i18n.G(`
+The find command queries the store for available packages.
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Oops :)

Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Please preserve the same style "Buys a snap" + "The buy command buys a snap from the store." or something.

File: cmd/snap/cmd_buy.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+var positiveResponse = map[string]bool{
+	"":            true,
+	i18n.G("y"):   true,
+	i18n.G("yes"): true,
+}
+
+type cmdBuy struct {
+	Currency  string `long:"currency" description:"ISO 4217 code for currency (https://en.wikipedia.org/wiki/ISO_4217)"`
+	Channel   string `long:"channel" description:"Use this channel instead of stable"`
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: I just realize this here.. do we buy from a channel?  We should be able to install from any channel after we buy, right?

File: cmd/snap/cmd_buy.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	BackendID string `long:"backend-id" description:"e.g. \"credit_card\", \"rest_paypal\""`
+	MethodID  int    `long:"method-id" description:"numeric identifier for a specific credit card or Paypal account"`
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Strange to have these here. What's the plan?  We don't expect normal users to say --method-id=42, right?

File: cmd/snap/cmd_buy.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	} `positional-args:"yes" required:"yes"`
+}
+
+func init() {
+	addCommand("buy", shortBuyHelp, longBuyHelp, func() flags.Commander {
+		return &cmdBuy{}
+	})
+}
+
+func (x *cmdBuy) Execute([]string) error {
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Please catch up with @mvo5 about how to error out if extra arguments are provided here.

File: cmd/snap/cmd_buy.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	snaps, resultInfo, err := cli.Find(&client.FindOptions{
+		Query: fmt.Sprintf("name:%s", opts.SnapName),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if len(snaps) < 1 {
+		return fmt.Errorf(i18n.G("cannot buy snap %q: it cannot be found"), opts.SnapName)
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: "cannot find snap %q"

File: cmd/snap/cmd_buy.go (hunk 3 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: This is in-case an app only exists in e.g. the unstable channel. The purchase is available in all channels.

File: cmd/snap/cmd_buy.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+	snap := snaps[0]
+
+	opts.SnapID = snap.ID
+
+	if opts.Channel == "" {
+		opts.Channel = snap.Channel
+	}
+
+	if opts.Currency == "" {
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Nipick: feels like a bit too much spacing here.  All the logic above is closely related, and could be joined together.

File: cmd/snap/cmd_buy.go (hunk 2 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Whoops! Will fix this

File: cmd/snap/cmd_buy.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		opts.Currency = resultInfo.SuggestedCurrency
+	}
+
+	opts.Price, opts.Currency, err = getPrice(snap.Prices, opts.Currency)
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot buy snap %q: it is free"), opts.SnapName)
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Assuming a generic error value means something specific is always a bad idea. This should be
```
("cannot buy snap %q: %v", opts.SnapName, err)
```

File: cmd/snap/cmd_buy.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	fmt.Fprintf(Stdout, i18n.G("Do you want to buy %q from %q for %s? (Y/n): "), snap.Name,
+		snap.Developer, formatPrice(opts.Price, opts.Currency))
+
+	response, _, err := reader.ReadLine()
+	if err != nil {
+		return err
+	}
+
+	if !positiveResponse[strings.ToLower(string(response))] {
+		return fmt.Errorf(i18n.G("buying snap %q cancelled by user"), opts.SnapName)
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Is it "you" or is it "the user"?  I think this can be simply "cancelled".

Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Or perhaps "aborting"

File: cmd/snap/cmd_buy.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	result, err := cli.Buy(opts)
+	if err != nil {
+		return err
+	}
+
+	// TODO Handle pay backends that require user interaction
+	fmt.Fprintf(Stdout, "Buy state: %s\n", result.State)
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: This is very machine-like. Please see the PR by @mvo5 fixing all command results.

File: cmd/snap/cmd_find.go [hunk 11]
```diff
@@ -34,15 +34,10 @@ var longFindHelp = i18n.G(`
...
+func getPrice(prices map[string]float64, currency string) (float64, string, error) {
 	// If there are no prices, then the snap is free
 	if len(prices) == 0 {
-		return ""
-	}
-
-	// If the snap is priced, but has been purchased
-	if status == "available" {
-		return i18n.G("bought")
+		return 0, "", fmt.Errorf("no prices found")
```
Comment (niemeyer) [inline cmd/snap/cmd_find.go]: It it "no prices found" or is it "snap is free"?  See earlier comment.

File: cmd/snap/cmd_find.go [hunk 12]
```diff
@@ -65,9 +60,29 @@ func getPrice(prices map[string]float64, currency, status string) string {
...
+func getPriceString(prices map[string]float64, suggestedCurrency, status string) string {
+	price, currency, err := getPrice(prices, suggestedCurrency)
+
+	// If there are no prices, then the snap is free
+	if err != nil {
+		return ""
+	}
+
+	// If the snap is priced, but has been purchased
+	if status == "available" {
```
Comment (niemeyer) [inline cmd/snap/cmd_find.go]: Before the comparison was != "priced".. which one is the right way?

**Comment by niemeyer** [comment]
Several comments, nothing too fundamental. Once these are handled, please ask for a re-review by @pedronis if I'm not around (back on Thursday).

File: cmd/snap/cmd_buy.go (hunk 3 above)
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: But that makes no sense, right?  Why should the user know which channel the snap is currently available in for being able to buy it?

**Comment by niemeyer** [comment]
Actually, please just wait for me. These UX issues need some discussion.

File: cmd/snap/cmd_buy.go (hunk 4 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: I guess I imagined it being used for scripting? Easy enough to remove, though, and probably will.

File: cmd/snap/cmd_buy.go (hunk 5 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Will do

File: cmd/snap/cmd_buy.go (hunk 3 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Will try removing the parameter and make a snap only available in one channel, and see if it all works.

Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Indeed, it should have been clear that Find doesn't take channel as a parameter, so we don't need it from the user.

File: cmd/snap/cmd_buy.go (hunk 4 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Removed

File: cmd/snap/cmd_find.go (hunk 12 above)
Comment (pete-woods) [inline cmd/snap/cmd_find.go]: I think == "available" is clearer, so will change the other instance to be the same

**Comment by pete-woods** [comment]
I have attempted to address all of these comments now. Thanks for the good catches! :)

File: cmd/snap/cmd_buy.go (hunk 4 above)
Comment (pete-woods) [inline cmd/snap/cmd_buy.go]: Just to be clear, in the next branch, I prompt the user from a list of pay methods with output like this:

Choice  Description
1)      **** **** **** 1111 (Visa, exp. 12/2030)
2)      PayPal
Select payment method: 1

File: cmd/snap/cmd_buy.go [hunk 13]
```diff
@@ -43,10 +43,7 @@ var positiveResponse = map[string]bool{
 }
 
 type cmdBuy struct {
-	Currency  string `long:"currency" description:"ISO 4217 code for currency (https://en.wikipedia.org/wiki/ISO_4217)"`
-	Channel   string `long:"channel" description:"Use this channel instead of stable"`
-	BackendID string `long:"backend-id" description:"e.g. \"credit_card\", \"rest_paypal\""`
-	MethodID  int    `long:"method-id" description:"numeric identifier for a specific credit card or Paypal account"`
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: \o/  :)

File: cmd/snap/cmd_buy.go [hunk 14]
```diff
@@ -95,21 +93,17 @@ func buySnap(opts *store.BuyOptions) error {
 	snap := snaps[0]
 
 	opts.SnapID = snap.ID
-
-	if opts.Channel == "" {
-		opts.Channel = snap.Channel
-	}
-
+	opts.Channel = snap.Channel
```
Comment (niemeyer) [inline cmd/snap/cmd_buy.go]: Still awkward to see this here. The user is not buying from a channel.

Not a blocker now that it's just an implementation detail, but we should fix it.

**Pull request merged**